algorithms/linfa-ftrl/src/lib.rs

```diff
@@ -0,0 +1,60 @@
+#![doc = include_str!("../README.md")]
+
+mod algorithm;
+mod error;
+mod hyperparams;
+
+use crate::hyperparams::FtrlValidParams;
+pub use algorithm::Result;
+pub use error::FtrlError;
+pub use hyperparams::FtrlParams;
+use linfa::Float;
+use ndarray::Array1;
+use ndarray_rand::RandomExt;
+use rand::distributions::Uniform;
+
+#[cfg_attr(
+    feature = "serde",
+    derive(Serialize, Deserialize),
+    serde(crate = "serde_crate")
+)]
+#[derive(Clone, Debug)]
+pub struct FTRL<F: Float> {
+    /// FTRL (Follow The Regularized Leader - proximal) is a linear model for CTR prediction in online learning settings.
+    /// It stores z and n values, which are later used to calculate weights at incremental model fit and during prediction.
+    /// It is a special type of linear model with sigmoid function which uses L1 and L2 regularization.
+    /// ```rust
+    /// use linfa::Dataset;
+    /// use ndarray::array;
+    /// use linfa_ftrl::FTRL;
+    /// use linfa::prelude::*;
+    /// let dataset = Dataset::new(array![[0.], [1.]], array![true, false]);
+    /// let params = FTRL::params();
+    /// let model = params.fit_with(None, &dataset).unwrap();
+    /// let predictions = model.predict(&dataset);
+    /// ```
+    params: FtrlValidParams<F>,
+    z: Array1<F>,
+    n: Array1<F>,
+}
+
+impl<F: Float> FTRL<F> {
+    /// Create a default parameter set for construction of Follow The Regularized Leader - proximal model
+    /// The description can be found here https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41159.pdf
+    ///
+    /// It requires data preprocessing done in the separate step.
+
+    /// Create default hyperparameters
+    pub fn params() -> FtrlParams<F> {
+        FtrlParams::default()
+    }
+
+    /// Create a new model with given parameters and number of features
+    pub fn new(params: FtrlValidParams<F>, nfeatures: usize) -> FTRL<F> {
+        Self {
+            params,
+            n: Array1::zeros(nfeatures),
+            z: Array1::random(nfeatures, Uniform::new(F::zero(), F::one())),
```

Comment:
`Array1::random` is problematic because it always uses `SmallRng` and is seeded by `ThreadRng`. For testing we want an RNG we seed ourselves for reproduceability and in the real world we want a higher-quality RNG like `Xoshiro` (assuming that the initial values of `z` actually matter for performance, if it doesn't then we might not even need to use an RNG here). Therefore I'd like this to call `Array1::random_using` instead with a custom RNG provided by `FtrlParams`. This means `FtrlParams` will need a type parameter of `R: Rng` and have a field for the RNG. This would also mean you can no longer clone the params directly into `FTRL`. In that case I suggest replicating the current fields of `FtrlParams` in `FTRL` and copying those fields instead.

Reply:
I tried to do this, but run into some problems. Maybe you could help me with them. Adding `R: Rng` to the `FtrlParams` by cascade causes in `fit_with` refer to `&mut self` instead of `&self`, because if we pass to `fit_with` `None` as model parameter, we must create default model and to do this we must use &mut params (and consequently &mut rng) meaning &mut self in `fit_with`. Maybe there is simple solution, it would be great if you can point me towards it.
On the other hand, `fit_with` IMHO is very important in online settings.

I would propose to use another approach. I would suggest to add to `FTRL::new(..)` parameter `rng: &mut dyn Rng`  to initialize `z` values. Additionally, add `get_rng()` to `FtrlParams` which returns `rand_xoshiro::Xoshiro256Plus`. Let me know what you think

Reply:
The way it's done in other parts of `linfa` is to clone the RNG when calling `fit_with` for the first time. This isn't as bad as it sounds because the clone is only done at the start. With your approach, I'm not sure how `fit_with(None)` would work, since we'd need to create a new RNG out of nowhere. `get_rng()` isn't good because `Xoshiro` doesn't necessarily fit every usecase, since sometimes a non-seeded RNG like `ThreadRng` is required (I misspoke in my comment about real world RNGs earlier). This is why we prefer having users provide their own RNGs.